module/zfs/dbuf.c

```diff
@@ -4560,10 +4569,12 @@ dbuf_lightweight_bp(dbuf_dirty_record_t *dr)
 		return (&dn->dn_phys->dn_blkptr[dr->dt.dll.dr_blkid]);
 	} else {
 		dmu_buf_impl_t *parent_db = dr->dr_parent->dr_dbuf;
+		ASSERT(MUTEX_HELD(&parent_db->db_mtx));
 		int epbs = dn->dn_indblkshift - SPA_BLKPTRSHIFT;
 		VERIFY3U(parent_db->db_level, ==, 1);
 		VERIFY3P(DB_DNODE(parent_db), ==, dn);
 		VERIFY3U(dr->dt.dll.dr_blkid >> epbs, ==, parent_db->db_blkid);
+		ASSERT(RW_LOCK_HELD(&parent_db->db_rwlock));
```

Comment:
We just return pointer here. Since indirect dbufs do nto relocate, we should not enforce the lock being held until the actual dereference point.